### DISCUSSION TITLE
resetApiState not working?

### DISCUSSION THREAD

**Discussion by Janaka-Steph** [reduxjs/redux-toolkit#2032] [Q&A] [answered]
reduxjs/toolkit: "1.7.2"

I tried hard to reset the result of a query but without success.

```
  const { data } = useIsReadyQuery();
  console.log('data);
  
  useEffect(() => {
    dispatch(walletUnlockerApi.util.resetApiState());
  }, []);
  ```
  
Here I expect `data` to be undefined because isReady is not supposed to work without some metadata I provide after. Instead it is returning cached data of last successful call.

I noticed this issue https://github.com/reduxjs/redux-toolkit/pull/1735 but it doesn't seem to be fixed. If the problem is my code I have no idea what I am doing wrong.

**Comment by phryneas** [comment, marked as answer]
can you provide a reproduction for this?

**Reply by Janaka-Steph** [reply]
Repo is public https://github.com/tdex-network/tdex-dashboard. I can make a branch with some `console.log` on `data`. But this alone will not work without setting up the server (through Docker) that `isReady` is calling. I don't know if you have time and interest to setup that.

**Reply by Janaka-Steph** [reply]
My current task is trying to call isReady on this page https://github.com/tdex-network/tdex-dashboard/blob/master/src/features/walletUnlocker/OnboardingPairing.tsx and not getting cached data.

**Reply by phryneas** [reply]
The feedback I *can* give from this: in `walletUnlocker.api.ts` you should be using `queryFn` instead of `query` and a `baseQuery` with a `switch..case` statement.

Also, you should probably have one big api with `injectEndoints` in all your separate api files and use `invalidateTags` to refetch parts of it. Not multiple apis. `resetApiState` is something meant to be done between tests, not really like that. If you need to get the real value after an `invalidate`, not the "last working" one, check for `currentData` instead of `data`.

But no, this is not working as a reproduction. I will need something small that I can just execute and observe.

**Reply by Janaka-Steph** [reply]
Thank you I will try to refactor `walletUnlocker.api.ts`.
I know I should use `injectEndoints` but I didn't manage splitting the APIs, was stuck on "ReferenceError: Cannot access uninitialized variable."
I checked `currentData` but it was equal to `data`.
Thank you for your time, I will dig more.

**Reply by phryneas** [reply]
If `currentData` is not resetting, your Redux state is not resetting at all. Can you see any change in the Redux devtools?

**Reply by Janaka-Steph** [reply]
I see the state is reset in Redux devtools yes

**Reply by phryneas** [reply]
So the action is dispatched and then all the cached data under that `reducerName` slice is removed?

That would be pretty much impossible, because `currentData` is always a value selected from the current Redux state, it is not cached anywhere in the component.

**Reply by Janaka-Steph** [reply]
The state I create myself are cleared properly. The state created by RTK (walletUnlockerService) is cleared, but calling `isReady` again will set the walletUnlockerService.queries.isReady to a cached value.

**Reply by phryneas** [reply]
That is literally impossible on the RTK Query side, that data is gone. At this point I would assume that your `client.isReady` function is doing some internal caching. I'm sorry, but without a *small* reproduction I don't think I can help you any further.

**Reply by Janaka-Steph** [reply]
Ok I will try to find out what is going on.

**Reply by Janaka-Steph** [reply]
@phryneas I am sorry for asking but I am trying to refactor walletUnlocker.api.ts as you suggested. I moved endpoints code into queryFn instead of query, but I don't understand what I should put into baseQuery param. 
If I put something like `baseQuery: fetchBaseQuery({ baseUrl: 'https://pokeapi.co/api/v2/' })` I got TS errors.
Do you mind giving me a hint? Thank you very much.

**Reply by Janaka-Steph** [reply]
The thing is I am using gRPC, so my understanding is that I should not need any baseQuery. Evrything is done in queryFn, like 
```
 genSeed: build.query<string[], void>({
      queryFn: async (arg, { getState }) => {
        try {
          const state = getState() as RootState;
          const client = selectWalletUnlockerClient(state);
          const metadata = selectMacaroonCreds(state);
          const genSeedReply = await client.genSeed(new GenSeedRequest(), metadata);
          return { data: genSeedReply.getSeedMnemonicList() };
        } catch (error) {
          console.error(error);
          return { error: (error as RpcError).message };
        }
      },
    })
```

**Reply by Janaka-Steph** [reply]
Ha I found it! `fakeBaseQuery<string>()` !
